src/impl_/extract_argument.rs

```diff
@@ -101,9 +101,21 @@ pub fn from_py_with_with_default<'py, T>(
 /// single string.)
 #[doc(hidden)]
 #[cold]
-pub fn argument_extraction_error(py: Python, arg_name: &str, error: PyErr) -> PyErr {
+pub fn argument_extraction_error(py: Python, arg_name: &str, mut error: PyErr) -> PyErr {
     if error.get_type(py) == PyTypeError::type_object(py) {
-        PyTypeError::new_err(format!("argument '{}': {}", arg_name, error.value(py)))
+        let mut reason = error
+            .value(py)
+            .str()
+            .unwrap_or_else(|_| PyString::new(py, ""))
+            .to_string();
+
+        while let Some(cause) = error.cause(py) {
+            reason.push_str(":\n\t");
+            reason.push_str(&cause.to_string());
+            error = cause
+        }
+
+        PyTypeError::new_err(format!("argument '{}': {}", arg_name, reason))
```

Comment:
I wonder, rather than flattening all the causes, should we be preserving the cause of the original error instead, like this:

```suggestion
        let remapped_error = PyTypeError::new_err(format!("argument '{}': {}", arg_name, error.value(py)));
        remapped_error.set_cause(py, error.cause(py));
        remapped_error
```

Then the error seen from Python might be more like this:

```
TypeError: 'str' object cannot be interpreted as an integer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/david/Dev/pyo3-scratch/test.py", line 19, in <module>
    pyo3_scratch.function_to_call_with_value_class(ValueClass("unparsable string"))
TypeError: argument 'value': failed to extract field ValueClass.value
```

I think preserving the cause tree like this is probably more correct. For example, the argument conversion code could run any old Python code (not just our PyO3 code) which generates some exception with its own separate cause. I think PyO3 flattening those causes into a single error message is a mistake; it might lose valuable debugging information.

Reply:
You're totally right with this remark. Also, I didn't completely liked the flattening so this is a better approach while still preserving all the relevant information. 

In addition, I tried to comply to rustfmt and clippy where relevant to this pull request. However, they were also marking parts that were untouched (some strict comparisons between floating points). Let's see if the workflows succeed.